pkg/discovery/summary.go

```diff
@@ -0,0 +1,238 @@
+/*
+Copyright the Sonobuoy contributors 2022
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package discovery
+
+import (
+	"os"
+	"path"
+	"path/filepath"
+	"io/fs"
+        "io/ioutil"
+	v1 "k8s.io/api/core/v1"
+	k8sver "k8s.io/apimachinery/pkg/version"
+	"github.com/sirupsen/logrus"
+	"encoding/json"
+)
+
+type ClusterSummary struct {
+	NodeHealth HealthInfo `json:"node_health" yaml:"node_health"`
+	PodHealth HealthInfo `json:"pod_health" yaml:"pod_health"`
+	APIVersion string `json:"api_version" yaml:"api_version"`
+	ErrorInfo LogSummary `json:"error_summary" yaml:"error_summary"`
+}
+
+
+type HealthInfo struct {
+	Total int `json:"total_nodes" yaml:"total_nodes"`
+	Healthy int `json:"healthy_nodes" yaml:"healthy_nodes"`
+	Details []HealthInfoDetails `json:"details,omitempty" yaml:"details,omitempty"`
+}
+
+type HealthInfoDetails struct {
+	Name string `json:"name" yaml:"name"`
+	Healthy bool `json:"healthy" yaml:"healthy"`
+	Ready string `json:"ready" yaml:"ready"`
+	Reason string `json:"reason,omitempty" yaml:"reason,omitempty"`
+	Message string `json:"message,omitempty" yaml:"message,omitempty"`
+	Namespace string `json:"namespace,omitempty" yaml:"namespace,omitempty"`
+}
+
+const (
+	//Filename of the core nodes json output, relative to ClusterResourceLocation
+	CoreNodesFile = "core_v1_nodes.json"
+
+	//Filename of the core pod json output, relative to NSResourceLocation
+	CorePodsFile = "core_v1_pods.json"
+
+	//Filename for the server version
+	ServerVersionFile = "serverversion.json"
+
+	//Filename of the cluster health, relative to MetaLocation
+	ClusterHealthFile = "clusterhealth.json"
+)
+
+// LoadJsonIntoStruct reads a json file into an object.
+// fileName needs to be the name of a file that can be opened
+// error will be returned if opening the file fails,
+// or if decoding its content as Json into the supplied "object" fails
+func LoadJsonIntoStruct(fileName string, object interface{}) (err error) {
```

Comment:
Is this the method you thought shouldn't be needed? Because you had an issue with an existing method? Could you refresh my memory on that?

Reply:
What I was doing was something like this: (this is a dummy example, I can go through the commits to see if i can find it if necessary)
```go
        file, _ := os.Open("test.txt")
        info, _ := file.Stat()
        _, ok := info.Sys().(io.Reader)
        if !ok {
                fmt.Println("Not a reader")
        }
```
This is just a simplification of the issue I was hitting, basically I was trying to open a file I wanted to process, and then I was passing it to `ExtractIntoStruct`, for example, and it was failing this test:

```go
                reader, ok := info.Sys().(io.Reader)
                if !ok {
                        return errors.New("info.Sys() is not a reader")
                }
```
Maybe it's my fault, and there's a better way to get a `os.FileInfo` object that has a `Sys()` that is a reader, I just couldn't find out, so I took the shortcut of creating `LoadJsonIntoStruct`

I'd say, as you said in a comment above, if we can make the `Reader` transparently work out of a tgz or an uncompressed directory, this issue would go away, and there would be no need to have a separate `LoadJsonIntoStruct`

Reply:
OK, lets just ignore the issue for the moment and consider it an issue for cleanup when that gets done then. 👍